framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -3495,55 +3489,49 @@ private void assertFunctionalInterface(
-            for (int i = 0; i < overriddenType.getTypeArguments().size(); i++) {
-                AnnotatedTypeMirror argType = overriddenType.getTypeArguments().get(i);
-                if (argType.getKind() == TypeKind.WILDCARD) {
-                    AnnotatedWildcardType wildcardType = (AnnotatedWildcardType) argType;
-
-                    final TypeMirror wildcardUbType =
-                            wildcardType.getExtendsBound().getUnderlyingType();
-                    final TypeMirror typeParamUbType =
-                            bounds.get(i).getUpperBound().getUnderlyingType();
-                    if (isExtendsWildcard(wildcardType)) {
-                        TypeMirror glbType =
-                                InternalUtils.greatestLowerBound(
-                                        this.checker.getProcessingEnvironment(),
-                                        typeParamUbType,
-                                        wildcardUbType);
-
-                        // checkTypeArgs now enforces that wildcard annotation bounds MUST be within
-                        // the bounds of the type parameter.  Therefore, the wildcard's upper bound
-                        // should ALWAYS be more specific than the upper bound of the type parameter
-                        // That said, the Java type does NOT have to be.
-                        // Add the annotations from the wildcard to the lub type.
-                        final AnnotatedTypeMirror newArg;
-                        if (types.isSameType(wildcardUbType, glbType)) {
-                            newArg = wildcardType.getExtendsBound().deepCopy();
-
-                        } else {
-                            newArg = this.toAnnotatedType(glbType, false);
-                            newArg.replaceAnnotations(
-                                    wildcardType.getExtendsBound().getAnnotations());
-                        }
-                        newTypeArguments.set(i, newArg);
+     * @param functionalType the functional interface type
+     * @param groundTargetJavaType the Java type as found by javac
+     */
+    private void makeGroundTargetType(
+            AnnotatedDeclaredType functionalType, DeclaredType groundTargetJavaType) {
+        if (functionalType.getTypeArguments().isEmpty()) {
+            return;
+        }
 
+        List<AnnotatedTypeMirror> newTypeArguments =
+                new ArrayList<>(functionalType.getTypeArguments());
+
+        for (int i = 0; i < functionalType.getTypeArguments().size(); i++) {
+            AnnotatedTypeMirror argType = functionalType.getTypeArguments().get(i);
+            if (argType.getKind() == TypeKind.WILDCARD) {
+                AnnotatedWildcardType wildcardType = (AnnotatedWildcardType) argType;
+
+                TypeMirror wildcardUbType = wildcardType.getExtendsBound().getUnderlyingType();
+                TypeMirror correctArgType = groundTargetJavaType.getTypeArguments().get(i);
```

Comment:
#1424 is caused by this line.
I don't see a check that `groundTagetJavaType` and `functionalType` have the same number of type arguments.
This is causing a crash somewhere else, too, but it's been harder to reproduce.